### ISSUE TITLE
Gradle 7.4+ Android SamplesVariantRule breaking with IDEA Gradle sync (Kotlin multiplatform)

### ISSUE THREAD

**Issue by sellmair** [gradle/gradle#20145]
Syncing a Kotlin Multiplatform Project with HMPP mode enabled used to work for 
- Gradle 7.0.2 ✅ 
- Gradle 7.1 ✅ 
- Gradle 7.2 ✅ 
- Gradle 7.3.3 ✅ 

But fails with 
- Gradle 7.4 ❌ 
- Gradle 7.4.1 ❌ 

### Current Behavior
The configuration "kotlinKlibCommonizerClasspath" cannot be resolved anymore during sync (CLI works fine)

```
Execution failed for task ':commonizeNativeDistribution'.
> Could not resolve all files for configuration ':kotlinKlibCommonizerClasspath'.
   > Could not find annotations-13.0-samplessources.jar (org.jetbrains:annotations:13.0).
     Searched in the following locations:
         file:/Users/sebastiansellmair/.m2/repository/org/jetbrains/annotations/13.0/annotations-13.0-samplessources.jar
   > Could not find trove4j-1.0.20200330-samplessources.jar (org.jetbrains.intellij.deps:trove4j:1.0.20200330).
     Searched in the following locations:
         file:/Users/sebastiansellmair/.m2/repository/org/jetbrains/intellij/deps/trove4j/1.0.20200330/trove4j-1.0.20200330-samplessources.jar
   > Could not find jna-5.6.0-samplessources.jar (net.java.dev.jna:jna:5.6.0).
     Searched in the following locations:
         file:/Users/sebastiansellmair/.m2/repository/net/java/dev/jna/jna/5.6.0/jna-5.6.0-samplessources.jar

```

### Context
- Requires IntelliJ with Android plugin applied (All recent versions affected)
- Requries Kotlin Multiplatform Project with HMPP enabled (sample project given below)

### Additional Resources: 
- Ticket on Kotlin side (to find a workaround) https://youtrack.jetbrains.com/issue/KT-51583
- Sample project 
https://github.com/sellmair/kotlin-multiplatform-projects/tree/KT-51583-Gradle_7.4%2B_SamplesVariantRule_interference_Could_not_resolve_all_files_for_configuration_kotlinKlibCommonizerClasspath

- Source Code of `SamplesVariantRule.kt`: 
https://github.com/JetBrains/android/blob/c5057a6bd25016ef9b4acd890c04ccd5476500c4/gradle-tooling/studio-gradle-tooling-impl/src/com/android/ide/gradle/model/artifacts/samples/SamplesVariantRule.kt

**Comment by sellmair** [comment]
cc @atyrin can you reproduce this?
@hegyibalint : I might assume that it could have something todo with my mavenLocal? Let me clear .m2 and try again!

**Comment by sellmair** [comment]
Deleted .m2 -> Imports successfully! ✅ 
Install kotlin master locally -> Imports successfully ✅ 
Install kotlin master locally using `./gradlew install -PdeployVersion=1.6.20-yazio` -> Issues reproduces ❌

**Comment by sellmair** [comment]
So in this case, there might be some interesting stuff going on, but it is not clear if just the publication is broken on my end with this setup. Until reported again from "real users", I think this issue can be classified as 'non-critical'?

**Comment by ljacomet** [comment]
Whenever Maven Local is at play, please take a look at https://docs.gradle.org/current/userguide/declaring_repositories.html#sec:case-for-maven-local

tl;dr: Maven local is a cache for maven but a real repo for Gradle. If a single file is missing from a module, Gradle considers it an error and does not look elsewhere.

Given that, I will close this issue.

**Comment by gavra0** [comment]
Using the attached project [repro.zip](https://github.com/gradle/gradle/files/8682420/repro.zip) and running `./gradlew mytask` fails with gradle 7.4.2, while it passes with 7.3.3. Note that the local Maven repo is valid, containing valid Maven metadata.

The issue seems to happen only with local Maven repos, as using `-SNAPSHOT` dependencies from AndroidX (see commented out lines in build.gradle) and `-dev` version for Kotlin tools does not reproduce the issue.

**Comment by gavra0** [comment]
Interestingly, adding at least one attribute to the configuration seems to fix the problem:
```
// Passes
configurations {
    myconf.defaultDependencies {
        it.add(project.dependencies.create('org.jetbrains.kotlin:kotlin-stdlib-common:1.7.255-SNAPSHOT'))
    }
    myconf {
        attributes {
            attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, "false-value"))
        }
    }
}
```
and this one fails (tested with AGP as we publish Gradle metadata)
``` 
configurations {
    myconf.defaultDependencies {
        it.add(project.dependencies.create('com.android.tools.build:gradle:7.3.0-alpha09')) // this one fails
    }
    myconf {
        attributes {
            attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, "false-value"))
        }
    }
}
```

**Comment by dsavvinov** [comment]
So, I dig around and got some interesting information about what was going on.


## Context 

The described issue is based upon following three "quirks"/facts, call them as you want:

1. In the core of Gradle dependency resolution (in `org.gradle.internal.component.model.ComponentAttributeMatcher#match`), `kotlin-klib-commonizer-embeddable` claims that it has no `consumableConfigurations` (explanation how it happened below)

2. `RealisedMavenModuleResolveMetadata` doesn't seem to add variants from `ComponentMetadataRule`s, while `DefaultMavenModuleResolveMetadata` does. 

3. In Gradle 7.3.3 `*ModuleResolveMetadata`s coming from `mavenLocal` were modelled as `RealisedMavenModuleResolveMetadata`, and in 7.4+ it changed to `DefaultMavenModuleResolveMetadata` (explanation what caused that change below)

## Failure explanation

Let's first demonstrate how those facts, when put together, lead to that undesired behaviour, and then proceed with discussion on why we even have those facts and whether it is desireable/intended. 

So, when Gradle Dependency Resolution doesn't find `consumableConfigurations`, it uses "fallback configuration", which is `default`. And due to Fact №1, indeed that was what were happening before Gradle 7.4, and one can observe that `kotlin-klib-commonizer-embeddable configuration=default` were chosen, and everything worked.

Now due to fact №3, internal Gradle representation of `kotlin-klib-commonizer-embeddable` has changed in 7.4, and just so it happens that due to Fact №2 this `ModuleResolveMetadata` now suddenly becomes aware of `ComponentMetadataRule`s and adds one `samplesssource` variant -- **to the empty list** of `consumableConfiguraiton`s!

Therefore, as that list is not empty anymore, fallback to `default` configuration doesn't happen, and the attribute matching runs against one `samplessource`-variant, rightfully chosing it (as there are no more options) and then failing as it is not present. 

> Note: we've actually tried to apply Ivan's workaround of requesting additional attributes, which were reasonable anyways, on the consumer side. It didn't help, but changed the error from `couldn't find X` to attribute mismatch: 
```
No matching variant of org.jetbrains.kotlin:kotlin-klib-commonizer-embeddable:1.7.255-SNAPSHOT was found. .
<...>
Variant 'samplessources' capability org.jetbrains.kotlin:kotlin-klib-commonizer-embeddable:1.7.255-SNAPSHOT:
    - Incompatible because this component declares attribute 'org.gradle.category' with value 'documentation' and the consumer needed attribute 'org.gradle.category' with value 'library'
```
> This should be pretty clear now: Gradle still choses the only one variant, but now, as we've added some attributes on the consumer, it even became incompatible. And it still the only one, so the only thing Dependency Resolution can do is to fail with attribute mismatch

## Facts explanation

### Why kotlin-klib-commonizer-embeddable doesn't have consumableConfigurations?

The thing is, Gradle seems to derive variants for variant-unaware libraries through so-called `VariantDerivationStrategy`. The most common one is `JvmEcosystemDerivationStrategy`, which, for example, creates one variant for `compile`-configuration and one for `runtime`-configuration, and some others.

For that strategy to work, either of two things should happen:
* The context of resolution is a project, and `JvmEcosystemPlugin` applied in that project 
* The context of resolution is script (makes sense I guess, given that Gradle scripts are JVM programs and live in JVM ecosystem)

Now, the quirk `kotlin-klib-commonizer-embeddable` has is that it is declared *in the root project*. Therefore, no variant derivation strategy applied there, and we get empty list of `variantsForGraphTraversal` a.k.a. `consumableConfigurations`! 

> Note: this also explains first half of Ivan's reproducer. As once can see, `build.gradle` indeed doesn't have any plugins applied. Applying `java`-plugin fixes the issue both for Ivan's reproducer, and for ours. 
> I were not able to reproduce the second half with attribute tricks though. It fails for me no matter which attributes I request on consumer side.

We had some reasons to do it, but now we're considering how we can change that, as we're not happy with the fragile resolution context we're putting ourselves in.

### Why behavior changed in Gradle 7.4?

I'm, obviously, can't speak about the idea behind that change, but I can point to the piece of code which did that. In the `org.gradle.api.internal.artifacts.dsl.DefaultComponentMetadataProcessor#processMetadata` condition for whether application of rules changed:

Gradle 7.3.3:
```
if (ia.getRules().isCacheable()) { <process with caching> }
```

Gradle 7.4:
```
if (shouldCacheComponentMetadataRule(ia, metadata)) { < process with caching> }

// ...

// declaration of `shouldCacheComponentMetadataRule`
boolean shouldCacheComponentMetadataRule(InstantiatingAction<ComponentMetadataContext> action, ModuleComponentResolveMetadata metadata) {
    return action.getRules().isCacheable() && metadata.isComponentMetadataRuleCachingEnabled();
}

// ...

// declaration of `isComponentMetadataRuleCachingEnabled
/*
* When set to false component metadata rules are not cached.
* Currently, we disable it just for local maven/ivy repository.
*
* Default value is true.
*/
boolean isComponentMetadataRuleCachingEnabled();
```

Basically, as you can see, since v. 7.4, Gradle doesn't apply caching for `ComponentMetadataRule`s for metadata from `mavenLocal`. Sounds even reasonable to me, I guess, as Gradle uses `mavenLocal` as local cache and can't track changes in it reliably.

However, it still sounds pretty far away from change in behaviour, right? Well, here's where my little knowledge of Gradle is not enough: on the cached-path "realisation" of `ModuleResolveMetadata` happens, and they are transformed into `RealisedMavenModuleResolveMetadata`. And on non-cached path such realisation doesn't happen, and `DefaultMavenModuleResolveMetadata` is produced. This is exactly Fact №3.

## Summary

The main action point is on our side to move `kotlinKlibCommonizerClasspath` from root project, and solve issues, which previously led to that decision, in some other way. We might need help there and will reach Gradle team accordingly. 

However, I'd like to know Gradle team's opinion on few interesting behaviours I've mentioned throught this investigation:

1. Is it intended that `RealisedMavenModuleResolveMetadata` and `DefaultMavenModuleResolveMetadata`  behavior is different w.r.t. `ComponentMetadataRule`s? Do you have plans on changing this?

2. Is it intended that cached and non-cached paths for application of `CMR` make different decisions about realisation? Do you have plans on changing this? 

The part about "changing" it is specifically important to the Kotlin team, because change in either 1 or 2 might cause older KGPs to start failing on non-mavenLocal dependencies as well, which would be a showstopper.

**Comment by h0tk3y** [comment]
@sellmair Are you aware of any work in progress on the fixes or workarounds in Kotlin and/or AGP? 
I want to find out when it would be safe to fix the inconsistency in the behavior of cached vs non-cached component metadata on the Gradle side, without deteriorating this issue.

**Comment by sellmair** [comment]
Hey Sergey! 
Do you think that a plugin like `kotlin-multiplatform` is allowed to perform 

```
rootProject.plugins.apply("jvm-ecosystem")
```

when applied to any project within the build?
Other than that, I think AGP team has no plans on that issue, right now.

**Comment by h0tk3y** [comment]
Hey @sellmair,
Thanks for the update!

Given that the configuration `kotlinKlibCommonizerClasspath` in the root project is created by the Kotlin plugin, I think it should be OK for the plugin to also apply the `jvm-ecosystem` plugin to the root project. That's the way to say that the plugin wants dependency resolution to follow the Gradle's built-in JVM-specific schemas & strategies. As of now, it won't conflict with any other schemas:
* The derivation strategy that was discussed above is applied through a non-public API, and we have only the JVM implementation.
* I would not expect dependencies of the root project to have mixed JVM and non-JVM values of the attributes that `jvm-ecosystem` configures the schema for.

On the other hand, configuring the root project (by either creating a configuration or applying the plugin to it) might not be a good idea in the context of [project isolation](https://gradle.github.io/configuration-cache/#project_isolation), as it would lead to coupling between all the projects that access the root project.

**Comment by sellmair** [comment]
Good point. Maybe we can schedule a Meeting to finally resolve this issue?
We potentially can avoid running this commonizer task in the rootProject all together?

The untraditional requirement for this task: 
This should only run once for the whole build, collecting all requirements of all multiplatform projects within this build. 

Are there any other options to perform such 'action'?

**Comment by h0tk3y** [comment]
> The untraditional requirement for this task:
> This should only run once for the whole build, collecting all requirements of all multiplatform projects within this build.
> 
> Are there any other options to perform such 'action'?

A [shared build service](https://docs.gradle.org/current/userguide/build_services.html) should likely be a good fit.

**Comment by gavra0** [comment]
Please do note that if this **must** run only once, build services do not have such guarantee ( see https://github.com/gradle/gradle/issues/18586). If this is only a performance optimization, and there could be multiple `kotlinKlibCommonizerClasspath`, build services are indeed a good fit.

**Comment by sellmair** [comment]
Proposal: 
Let's go with a two step plan: 

Step 1: 
As long as we did not implement this commonizer task as shared build service, we already break project isolation. Therefore we can apply the java-base plugin also. This can be shipped in 1.8.0 and will unblock this issue. 

Step 2:
Implement this as shared build service in Kotlin 1.8.20 (or 1.9.0). 

Is this okay for everyone?

- [ ] @h0tk3y 
- [ ] @dsavvinov 
- [ ] @gavra0

**Comment by dsavvinov** [comment]
Seems like an acceptable option for now 👍

**Comment by h0tk3y** [comment]
> Therefore we can apply the java-base plugin also

You mean the `jvm-ecosystem` plugin, right? LGTM, then.

**Comment by sellmair** [comment]
Since everyone agree: 
I will create post the commit in kotlin.git/master and the corresponding YT issue here tomorrow. 
You can expect this to be in Kotlin 1.8

**Comment by sellmair** [comment]
Ticket to track implementing the task as build service: 
https://youtrack.jetbrains.com/issue/KT-53527/Implement-NativeDistributionCommonizerTask-as-shared-build-service

**Comment by sellmair** [comment]
Pushed the corresponding fix to master. 
This will be shipped in Kotlin 1.8

Thank you folks, a lot. 
Quite a hard one! Let's hope for the best 👍

**Comment by big-guy** [comment]
Let's verify this has been handled and close this issue.

**Comment by h0tk3y** [comment]
@big-guy Please note that the change that was expected in Gradle, https://github.com/gradle/gradle/pull/21213, has not been merged.

**Comment by ghale** [comment]
@h0tk3y My understanding from parsing through the conversation above and playing around with the reproducer is the following:
- The issue caused by the change in behavior between Gradle 7.3 and 7.4 was fixed in Kotlin 1.8
- Your PR fixes the inconsistent behavior between maven local and external dependencies, but we chose not to merge until the afore mentioned fix to Kotlin was in place because it would have caused external dependencies to exhibit the same behavior.
- Now that the issue is fixed on the Kotlin side, we are safe to merge your PR.

Is that understanding correct?
